File: src/main/java/ru/ioleksiv/telegram/bot/api/model/method/group/SetChatPhoto.java
package ru.ioleksiv.telegram.bot.api.model.method.group;

import org.springframework.core.io.FileSystemResource;
import ru.ioleksiv.telegram.bot.api.model.NetworkError;
import ru.ioleksiv.telegram.bot.core.controller.network.FileNetworker;
import ru.ioleksiv.telegram.bot.core.model.method.UploadFile;
import ru.ioleksiv.telegram.bot.core.model.responses.CommonResponse;
import ru.ioleksiv.telegram.bot.core.model.responses.ResponseCollection;

/**
 * @see <a href="https://core.telegram.org/bots/api#setchatphoto">setChatPhoto</a>
 */
public class SetChatPhoto extends UploadFile<Boolean> {
    private static final String METHOD = "setChatPhoto";

    /**
     * chat_id Integer or String Unique identifier for the target chat or username of the
     * target channel (in the format @channelusername)
     */
    private static final String CHAT_ID_KEY = "chat_id";
    /**
     * photo InputFile New chat photo, uploaded using multipart/form-data
     */
    private static final String PHOTO_KEY = "photo";

    public SetChatPhoto(FileNetworker fileNetworker) {
        super(METHOD, fileNetworker);
    }

    @Override
    protected Class<? extends CommonResponse<Boolean>> getResultWrapperClass() {
        return ResponseCollection.BooleanResponse.class;
    }

    public SetChatPhoto setPhoto(FileSystemResource photo) {
        putFile(PHOTO_KEY, photo);
        return this;
    }

    public SetChatPhoto setChatId(Long chatId) {
        putLong(CHAT_ID_KEY, chatId);
        return this;
    }

    @Override
    public SetChatPhoto setNetworkErrorListener(NetworkError onNetworkError) {
        pSetNetworkErrorListener(onNetworkError);
        return this;
    }
}

File: server-api/src/main/kotlin/io/github/oleksivio/tl/kbot/server/api/method/group/SetChatPhoto.kt
package io.github.oleksivio.tl.kbot.server.api.method.group

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import io.github.oleksivio.tl.kbot.server.api.model.ApiDict
import io.github.oleksivio.tl.kbot.server.api.model.BooleanResponse
import io.github.oleksivio.tl.kbot.server.api.model.method.Action
import io.github.oleksivio.tl.kbot.server.api.objects.ChatId
import io.github.oleksivio.tl.kbot.server.api.objects.InputFile

/**
 * @see [setChatPhoto](https://core.telegram.org/bots/api/#setchatphoto)
 */
data class SetChatPhoto(

    /**
     * chat_id Integer or String Unique identifier for the target chat or username of the
     * target channel (in the format @channelusername)
     */
    @JsonProperty(ApiDict.CHAT_ID_KEY)
    val chatId: ChatId,
    /**
     * photo InputFile New chat photo, uploaded using multipart/form-data
     */
    @JsonProperty(ApiDict.PHOTO_KEY)
    val photo: InputFile

) : Action<Boolean>() {

    @JsonIgnore
    override val resultWrapperClass = BooleanResponse::class

    @JsonProperty(ApiDict.METHOD_KEY)
    override val method = "setChatPhoto"
}
